tw crps tutorial

nicholasloveday opened the pull request:
## Tutorial notebook 
- [x] Short introduction to why you would use that metric and what it tells you
- [x] A link to a reference
- [x] A "things to try next" section at the end
- [x] Add notebook to [Tutorial_Gallery.ipynb](https://github.com/nci/scores/blob/develop/tutorials/Tutorial_Gallery.ipynb)
- [x] Optional - a detailed discussion of how the metric works at the end of the notebook

tennlee commented:
@nicholasloveday  I have just pushed this to 228, it will take 5 minutes or so to build

nicholasloveday commented:
Thanks @tennlee . It notebook is now up here https://scores.readthedocs.io/en/228-documentation-testing-branch/tutorials/Threshold_Weighted_CRPS_for_Ensembles.html

DerynGriffiths commented:
@nicholasloveday this will take a while to work through and comment on. Some minor typos I noticed in the meantime:

Rather than start a sentence as " twCRPS is defined as..." use "The definition of twCRPS is..." for nicer formatting.

"For evaluation ensembles, ..." should probably be "For evaluation of ensembles, ..."  

Replace "which will should take" with "which will take." or "which should take"

DerynGriffiths commented:
Some thoughts as I go. Apologies if I later contradict any of these thoughts. 

1. In this and other tutorial notebooks there is sometimes code for the sake of plotting explanatory graphs. If we could instead render the figure without the code (e.g. .jpg) we could remove code to explain the concepts and leave just code to explain how to use scores. This might make some of the notebooks less daunting. It is particularly relevant to this notebook as it is quite long.

2. In the "CRPS for Ensembles" tutorial add a link to this tutorial for threshold weighting.

3. In the very first paragraph of this tutorial note that the crps_cdf function has a threshold_weight functionality, but if using crps_for_ensembles there is no such functionality, and you will instead need to use one of the tw_crps_for_ensemble functions illustrated in this tutorial.

4. In the "Overview and graphical illustration of twCRPS" section I suggest starting with the graphical illustration of a step function weighting function and the area under the graph. I don't think you need to make it an ensemble forecast for this. If you do make it an ensemble you might like to make it a step function rather than joining ensemble members linearly. Please consider. Possibly make the mathematical definition a separate section. It is important in defining the chaining function. Consider leaving out the continuous version and jumping straight to the ensemble version as the relevant one here. You will still need to plot an example of v(x) for w(x).

5. Consider showing the step functions for tail_tw_crps_for_ensemble and interval_tail_tw_crps_for_ensemble in their own section with a simple use of the functions using dummy data and matching either the example from the crps_for_ensembles tutorial or the example used in the overview. This would separate out "how to use" from the more complicated examples with realistic data. Rename the real data section as "Calculating twCRPS with real data"

6. Only use one example of real data (e.g. Wind) to cut down the complexity.  Separate out extracting the data and discussing weighting the grid cells by latitude from using the tw_crps function. Reduce brain load on the reader by naming era5_wind as era5_observation, ifs_ens_wind as ifs_ens_forecast and neural_gcm_wind as neural_gcm_forecast.

7. Only use one of ifs_ens_wind and neural_gcm_wind. Add the option to explore the other as one to the "things to try next". 

8. And/or, add the option to compare ifs_ens_wind to neural_gcm_wind using unweighted (or constant weighted) crps

9. Add an example of v(x) for a trapezoidal weighting function. E.g v(x) = 0 for x < 0, x^2/2 for x between 0 and 1, and x for x > 1. Or give as an example of v(x) with an exercise to determine the corresponding weighting function.

10. If you want to provide an example of extracting data, maybe that should be another tutorial, and it could bring together a few scoring functions. 

11. I like the last example because you don't need to extract any more data. It makes it cleaner.

12. Overall I feel it is too long, with too much that is an aside to the main functions being demonstrated.

13. I like the reminder in the other notebooks to read the doc string via tail_tw_crps_for_ensemble? and similar. These can be commented out.

tleeuwenburg commented:
I concur with most but not all of Deryn's feedback items. Some of my feedback is actually a bit inconsistent with my normal views, and I'll try to explain that as I go. I'm happy for you to just take it into account and proceed as you see fit.

(1) I quite like having all the plotting code there, it's instructive for people who want to do the same kind of thing.

(12) This is one of the most complete end-to-end examples, which makes it a bit special. It's doing a kind of double-duty, both as an explanation of the tw_crps score, but also as a complete example showing a realistic use case which compared models, as well as demonstrating the use of Xarray's support for cloud-hosted Zarr datasets. There is a lot going on, and this time I think it's in a good way. A complicated response to this situation would be to fork out another kind of tutorial which isn't focused on explaining a score, but on using several scores to analyse two models in a more realistic way, while simplifying the tw_crps notebook to thin out the complexity and focus on the presentation of the score. However, I'm not sure that's worth the trouble compared to retaining both the length and complexity of this notebook, and using it as an exemplar for anyone seeking to understand more about intercomparison of models.

(13) I agree, please add the commented-out help reminders.

(14) In addition to the points made in (12), this example also examines a machine learning model (aka data-driven model, aka neural earth system model). This also makes it special. There is a lot of interest in the evaluation of ML models. The notebook as fine as-is, but is there any particular reason that this score was chosen for an ML model? i.e. is tw_crps particularly important for ML model analysis, or is it just as relevant for physical model intercomparison?

In general, I really like this notebook, and my general leaning is not to reduce complexity even though that's what I'd normally suggest.

tennlee commented:
I concur with most but not all of Deryn's feedback items. Some of my feedback is actually a bit inconsistent with my normal views, and I'll try to explain that as I go. I'm happy for you to just take it into account and proceed as you see fit.

(1) I quite like having all the plotting code there, it's instructive for people who want to do the same kind of thing.

(12) This is one of the most complete end-to-end examples, which makes it a bit special. It's doing a kind of double-duty, both as an explanation of the tw_crps score, but also as a complete example showing a realistic use case which compared models, as well as demonstrating the use of Xarray's support for cloud-hosted Zarr datasets. There is a lot going on, and this time I think it's in a good way. A complicated response to this situation would be to fork out another kind of tutorial which isn't focused on explaining a score, but on using several scores to analyse two models in a more realistic way, while simplifying the tw_crps notebook to thin out the complexity and focus on the presentation of the score. However, I'm not sure that's worth the trouble compared to retaining both the length and complexity of this notebook, and using it as an exemplar for anyone seeking to understand more about intercomparison of models.

(13) I agree, please add the commented-out help reminders.

(14) In addition to the points made in (12), this example also examines a machine learning model (aka data-driven model, aka neural earth system model). This also makes it special. There is a lot of interest in the evaluation of ML models. The notebook as fine as-is, but is there any particular reason that this score was chosen for an ML model? i.e. is tw_crps particularly important for ML model analysis, or is it just as relevant for physical model intercomparison?

In general, I really like this notebook, and my general leaning is not to reduce complexity even though that's what I'd normally suggest.

nicholasloveday commented:
Thanks @DerynGriffiths and @tennlee. 
Regarding complexity, I have embedded simple examples in the docs already for each twCRPS function (e.g., see the example at the end of https://scores.readthedocs.io/en/stable/api.html#scores.probability.tail_tw_crps_for_ensemble). I see the tutorial notebooks as an opportunity to explore more than just the basics of the function as that information should already all be in the docs.

1. Left these in the notebook
2. Done
3. Done
4a. Updated the CDF plot to "step" rather than linearly interpolate. Thanks for this pick up @DerynGriffiths!
4b. Removed first equation and now it jumps straight into the ensemble equation
5. Did not update. I think this is covered in the docs already
6. Left as is since @tennlee appreciated this tutorial being more comprehensive
7. Left as is, since physical vs ML models is of great interest to the community at the moment.
8. Added into the things to try next section
9. Replaced example 3 with your suggestion. Please check this @DerynGriffiths . I included a plot of v(x) and w(x). It turns out that in this case v(x) is the Huber Loss function which is kind of nice.
10. I think a dedicated tutorial for pulling in data from other sources would be good. This can be done seperately
13. Done
14. Verification of extremes is commonly highlighted as important to do with data driven models. I've never seen anyone evaluate them using threshold weighted scores before. I added a sentence about this. Also note that Neural GCM is a hybrid model rather than a pure data driven model.

Other changes. 
- Added plots of v(x) and w(x) for the interval weighting
- Made plots narrower to hopefully fit in the docs better

I'll push up the changes shortly.

tennlee commented:
As far as I can tell all feedback has been addressed, and the notebook looks great. I'm going to proceed with the merge, but feel free to continue discussions and raise additional issues for any further improvements that anyone would still like to see made.
